kv/dist_sender.go

```diff
@@ -520,6 +526,50 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
 	return desc, descNext, nil
 }
 
+// getDescriptorsForReverseScan takes a call and looks up the corresponding range
+// descriptors associated with it. First, the range descriptor for
+// call.Args.EndKey is looked up. If call.Args.Key less than the
+// returned descriptor, the next descriptor is obtained as well.
+func (ds *DistSender) getDescriptorsForReverseScan(call proto.Call, options lookupOptions) (*proto.RangeDescriptor, *proto.RangeDescriptor, error) {
+	options.useReverseScan = true
+	var desc *proto.RangeDescriptor
+	var err error
+
+	revScanRequest := call.Args.(*proto.ReverseScanRequest)
+	// If the Args.EndKey is the range.StartKey,if include EndKey,we get range descriptor throug
+	// Args.EndKey,else used the EndKey.Prev() to get previous range desciptor.
+	if revScanRequest.InclusiveEndKey {
+		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().EndKey, options)
+	} else {
+		desc, err = ds.rangeCache.LookupRangeDescriptor(call.Args.Header().EndKey.Prev(), options)
```

Comment:
I'm very interested in getting rid of the `Prev()` here. The `rangeCache` is a tree, it has both `Ceil()` and `Floor()` so the only problem is that `Floor()` gives you the wrong thing for a key which is in the tree (because it stops on equality). But you can wrap the key in a way that makes it never compare equal to anything else, and then flooring on that should give you the desired previous descriptor (if cached, but you can check easily if what you get back matches).
That's a bit of complication but can probably be wrapped up nicely. I just feel that reintroducing `Prev()` causes a lot of tricky issues.